File: src/configs.ts
import * as fs from "fs";
import * as path from "path";
import * as os from "os";

export type ConfigsProps = {
    arduino_cli_path: string;
    arduino_data_path: string;
    arduino_downloads_path: string;
    arduino_sketch_path: string;
    auto_start: boolean;
    // preferences.txt
    arduino_preferences_path: string;
    arduino_sketch_path_from_preferences: boolean;
    additional_urls_from_preferences: boolean;
};

const arduin_cli_name: Record<string, string> = {
    darwin: "arduino-cli",
    linux: "arduino-cli-ubuntu-x64",
    win32: "arduino-cli.exe",
};

export const CONFIG_DIR = path.join(os.homedir(), "AppData", "Local", "FlowcodeAgent");
export const CONFIG_FILE = path.join(CONFIG_DIR, "configs.json");

const DEFAULT_PREFS_PATH = path.join(os.homedir(), "AppData", "Local", "Arduino15", "preferences.txt");

function read_prefs(prefsPath: string): Record<string, string> {
    try {
        if (fs.existsSync(prefsPath)) {
            const content = fs.readFileSync(prefsPath, "utf8");
            const result: Record<string, string> = {};
            for (const line of content.split(/\r?\n/)) {
                const eq = line.indexOf("=");
                if (eq > 0) result[line.slice(0, eq).trim()] = line.slice(eq + 1).trim();
            }
            return result;
        }
    } catch (err) {
        console.error("Error reading preferences.txt:", err);
    }
    return {};
}

const DEFAULTS: ConfigsProps = {
    arduino_cli_path: path.normalize(
        os.homedir() + "/AppData/Local/Programs/Arduino IDE/resources/app/lib/backend/resources/" + (arduin_cli_name[os.platform()] ?? "arduino-cli")
    ),
    arduino_data_path: path.join(os.homedir(), "AppData", "Local", "Arduino15"),
    arduino_downloads_path: path.join(os.homedir(), "AppData", "Local", "Arduino15", "staging"),
    arduino_sketch_path: path.join(os.homedir(), "Documents", "Arduino"),
    auto_start: true,
    arduino_preferences_path: DEFAULT_PREFS_PATH,
    arduino_sketch_path_from_preferences: true,
    additional_urls_from_preferences: true,
};

let _configs: ConfigsProps = { ...DEFAULTS };

export function load_configs(): ConfigsProps {
    try {
        if (!fs.existsSync(CONFIG_DIR)) {
            fs.mkdirSync(CONFIG_DIR, { recursive: true });
        }
        if (!fs.existsSync(CONFIG_FILE)) {
            fs.writeFileSync(CONFIG_FILE, JSON.stringify(DEFAULTS, null, 2), "utf8");
            _configs = { ...DEFAULTS };
        } else {
            const raw = fs.readFileSync(CONFIG_FILE, "utf8");
            _configs = { ...DEFAULTS, ...JSON.parse(raw) };
        }

        // override arduino_sketch_path from preferences.txt if enabled
        if (_configs.arduino_sketch_path_from_preferences) {
            const prefs = read_prefs(_configs.arduino_preferences_path);
            if (prefs["sketchbook.path"]) {
                _configs.arduino_sketch_path = prefs["sketchbook.path"];
            }
        }
    } catch (err) {
        console.error("Failed to load configs, using defaults:", err);
        _configs = { ...DEFAULTS };
    }
    return _configs;
}

export function get_configs(): ConfigsProps {
    return _configs;
}

/** อ่าน additional_urls จาก preferences.txt (ถ้าเปิดใช้งาน) */
export function get_additional_urls_from_preferences(): string[] {
    if (!_configs.additional_urls_from_preferences) return [];
    const prefs = read_prefs(_configs.arduino_preferences_path);
    const raw = prefs["boardsmanager.additional.urls"];
    if (!raw) return [];
    return raw.split(",").map(u => u.trim()).filter(Boolean);
}

File: src/configs.rs
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::sync::{OnceLock, RwLock};

use serde::{Deserialize, Serialize};

// ── Config struct ──────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Configs {
    pub arduino_cli_path: String,
    pub arduino_data_path: String,
    pub arduino_downloads_path: String,
    pub arduino_sketch_path: String,
    pub auto_start: bool,
    pub arduino_preferences_path: String,
    pub arduino_sketch_path_from_preferences: bool,
    pub additional_urls_from_preferences: bool,
}

fn arduino_cli_name() -> &'static str {
    if cfg!(target_os = "windows") {
        "arduino-cli.exe"
    } else if cfg!(target_os = "macos") {
        "arduino-cli"
    } else {
        "arduino-cli-ubuntu-x64"
    }
}

pub fn config_dir() -> PathBuf {
    dirs::home_dir()
        .expect("no home dir")
        .join("AppData")
        .join("Local")
        .join("FlowcodeAgent")
}

pub fn config_file() -> PathBuf {
    config_dir().join("configs.json")
}

impl Default for Configs {
    fn default() -> Self {
        let home = dirs::home_dir().expect("no home dir");
        let arduino15 = home.join("AppData").join("Local").join("Arduino15");
        Self {
            arduino_cli_path: home
                .join("AppData")
                .join("Local")
                .join("Programs")
                .join("Arduino IDE")
                .join("resources")
                .join("app")
                .join("lib")
                .join("backend")
                .join("resources")
                .join(arduino_cli_name())
                .to_string_lossy()
                .to_string(),
            arduino_data_path: arduino15.to_string_lossy().to_string(),
            arduino_downloads_path: arduino15.join("staging").to_string_lossy().to_string(),
            arduino_sketch_path: home
                .join("Documents")
                .join("Arduino")
                .to_string_lossy()
                .to_string(),
            auto_start: true,
            arduino_preferences_path: arduino15
                .join("preferences.txt")
                .to_string_lossy()
                .to_string(),
            arduino_sketch_path_from_preferences: true,
            additional_urls_from_preferences: true,
        }
    }
}

// ── preferences.txt parser ─────────────────────────────────────────────────

fn read_prefs(path: &str) -> HashMap<String, String> {
    let mut map = HashMap::new();
    if let Ok(content) = fs::read_to_string(path) {
        for line in content.lines() {
            if let Some(eq) = line.find('=') {
                let k = line[..eq].trim().to_string();
                let v = line[eq + 1..].trim().to_string();
                map.insert(k, v);
            }
        }
    }
    map
}

// ── global config store ────────────────────────────────────────────────────

static CONFIGS: OnceLock<RwLock<Configs>> = OnceLock::new();

fn configs_lock() -> &'static RwLock<Configs> {
    CONFIGS.get_or_init(|| RwLock::new(Configs::default()))
}

pub fn load_configs() {
    let dir = config_dir();
    let file = config_file();

    if !dir.exists() {
        fs::create_dir_all(&dir).ok();
    }

    let mut configs: Configs = if file.exists() {
        let raw = fs::read_to_string(&file).unwrap_or_default();
        serde_json::from_str(&raw).unwrap_or_default()
    } else {
        let defaults = Configs::default();
        if let Ok(json) = serde_json::to_string_pretty(&defaults) {
            fs::write(&file, json).ok();
        }
        defaults
    };

    if configs.arduino_sketch_path_from_preferences {
        let prefs = read_prefs(&configs.arduino_preferences_path);
        if let Some(p) = prefs.get("sketchbook.path") {
            configs.arduino_sketch_path = p.clone();
        }
    }

    *configs_lock().write().unwrap() = configs;
}

pub fn get_configs() -> Configs {
    configs_lock().read().unwrap().clone()
}

pub fn get_additional_urls_from_preferences() -> Vec<String> {
    let configs = get_configs();
    if !configs.additional_urls_from_preferences {
        return vec![];
    }
    let prefs = read_prefs(&configs.arduino_preferences_path);
    prefs
        .get("boardsmanager.additional.urls")
        .map(|s| {
            s.split(',')
                .map(|u| u.trim().to_string())
                .filter(|s| !s.is_empty())
                .collect()
        })
        .unwrap_or_default()
}
